lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/RestRequestExecutor.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Org.Apache.REEF.Client.Yarn.RestClient.v1;
+using Org.Apache.REEF.Tang.Annotations;
+using RestSharp;
+
+namespace Org.Apache.REEF.Client.Yarn.RestClient
+{
+    [DefaultImplementation(typeof(RestRequestExecutor))]
+    internal interface IRestRequestExecutor
+    {
+        Task<T> ExecuteAsync<T>(
+            IRestRequest request,
+            Uri uri,
+            CancellationToken cancellationToken) where T : new();
+    }
+
+    internal class RestRequestExecutor : IRestRequestExecutor
+    {
+        private readonly IRestClientFactory _clientFactory;
+
+        [Inject]
+        private RestRequestExecutor(
+            IRestClientFactory clientFactory)
+        {
+            _clientFactory = clientFactory;
+        }
+
+        public async Task<T> ExecuteAsync<T>(
+            IRestRequest request,
+            Uri requestUri,
+            CancellationToken cancellationToken) where T : new()
+        {
+            var client = _clientFactory.CreateRestClient(requestUri);
+
+            var response =
+                await
+                    client.ExecuteTaskAsync<T>(request, cancellationToken);
+            if (response.ErrorException != null)
+            {
+                throw new Exception("Executing REST API failed", response.ErrorException);
+            }
+
+            if (response.StatusCode != HttpStatusCode.OK)
+            {
+                var errorResponse = JsonConvert.DeserializeObject<Error>(response.Content);
```

Comment:
I've noticed that sometimes the RM returns XML instead of JSON. You may want to be careful with this.